src/group_theory/dihedral.lean

```diff
@@ -0,0 +1,391 @@
+
+instance : group (infinite_dihedral) :=
+begin
+  refine_struct
+  { .. infinite_dihedral.has_one,
+    .. infinite_dihedral.has_mul,
+    .. infinite_dihedral.has_inv };
+  repeat {intro g, cases g};
+  simp only
+   [one_eq, r_inv, s_inv, rr_mul, rs_mul, sr_mul, ss_mul,
+    zero_add, add_zero, sub_zero, add_neg, sub_self, neg_add_self,
+    add_assoc, add_sub_assoc, sub_add, sub_sub]
+end
+
+instance inc_hom : is_monoid_hom inc :=
+{ map_one := rfl,
+  map_mul := λ a b, by { cases a, cases b,
+    rw [infinite_cyclic.rr_mul, inc, inc, inc, rr_mul]}}
+
+section hom_from_gens
+
+variables {G : Type*} [group G] (r0 : G) {s0: G}
+variables (hs : s0 * s0 = 1)
+          (hrs : r0 * s0 * r0 = s0)
+
+include r0 s0 hs hrs
+
+def hom_from_gens :
+ (infinite_dihedral) → G
+| (r i) := r0 ^ i
+| (s i) := r0 ^ i * s0
+
+lemma hom_from_gens_r (i : ℤ) :
+hom_from_gens r0 hs hrs (r i) = r0 ^ i := rfl
+
+lemma hom_from_gens_s (i : ℤ) :
+hom_from_gens r0 hs hrs (s i) = r0 ^ i * s0 := rfl
+
+lemma sr_rs : ∀ (i : ℤ), r0 ^ i * s0 * (r0 ^ i) = s0 :=
+begin
+  have h_nat : ∀ (i : ℕ), r0 ^ i * s0 * (r0 ^ i) = s0,
+  begin
+    intro i,
+    induction i with i ih,
+    {rw [pow_zero, one_mul, mul_one]},
+    exact calc
+      r0 ^ (i + 1) * s0 * r0 ^ (i + 1) =
+        r0 ^ (i + 1) * s0 * r0 ^ (1 + i) : by rw [add_comm i 1]
+      ... = (r0 ^ i * r0) * s0 * (r0 * r0 ^ i) : by rw [pow_add, pow_add, pow_one]
+      ... = (r0 ^ i) * ((r0 * s0) * (r0 * (r0 ^ i))) :
+        by rw [mul_assoc (r0 ^ i) r0 s0, mul_assoc (r0 ^ i)]
+      ... = (r0 ^ i) * (((r0 * s0) * r0) * r0 ^ i) :
+        by rw [mul_assoc (r0 * s0) r0 (r0 ^ i)]
+      ... = s0 : by rw [hrs, ← mul_assoc, ih]
+  end,
+  rintro (i | i),
+  { exact h_nat i},
+  { dsimp [gpow],
+    rw [mul_inv_eq_iff_eq_mul, inv_mul_eq_iff_eq_mul, ← mul_assoc],
+    exact (h_nat (i + 1)).symm }
+end
+
+lemma sr_rs' (i : ℤ) : s0 * (r0 ^ i) = (r0 ^ (- i)) * s0 :=
+calc
+  s0 * r0 ^ i = 1 * (s0 * r0 ^ i) : (one_mul _).symm
+  ... = (r0 ^ ((- i) + i))  * (s0 * r0 ^ i) :
+     by {rw [← pow_zero r0, neg_add_self], refl}
+  ... = (r0 ^ (- i)) * ((r0 ^ i) * s0 * (r0 ^ i)) :
+     by {rw [gpow_add, mul_assoc (r0 ^ (- i)), mul_assoc]}
+  ... = r0 ^ (- i) * s0 : by rw [sr_rs r0 hs hrs i]
+
+instance is_hom_from_gens : is_monoid_hom (hom_from_gens r0 hs hrs) :=
+let h := sr_rs' r0 hs hrs in
+{ map_one := by {rw [one_eq, hom_from_gens_r], refl},
+  map_mul := by
+   { intros a b,
+     cases a with i i; cases b with j j;
+     simp [rr_mul, rs_mul, sr_mul, ss_mul, hom_from_gens_r, hom_from_gens_s],
+     {rw [gpow_add]},
+     {rw [← mul_assoc, gpow_add]},
+     {rw [mul_assoc, h j, ← mul_assoc, gpow_add]},
+     {rw [mul_assoc, ← mul_assoc s0 _ s0, h j, mul_assoc, hs, mul_one, gpow_add]}}}
+
+end hom_from_gens
+
+section monoid_hom_from_gens
+
+variables {M : Type*} [monoid M] (r0 : M) {s0: M}
+variables (hs : s0 * s0 = 1)
+          (hrs : r0 * s0 * r0 = s0)
+
+include r0 s0 hs hrs
+
+def r0_unit : units M :=
```

Comment:
This name isn't very descriptive. Names shouldn't usually depend on the names of variables. Maybe this is more of a local definition so it's okay.